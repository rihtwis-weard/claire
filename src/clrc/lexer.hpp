#pragma once

#include <vector>

#include "state_machine.h"
#include "token.hpp"

namespace claire {
  class Lexer {
    LexicalState       state_;
    std::string        source_;
    std::vector<Token> tokens_;

  public:
    explicit Lexer(std::string const &fpath)
      : state_{LexicalState::eNextChar}
      , source_{loads(fpath)} {
    }

    auto lex() {
      Token tok{};

      // Alias pointer to buffer to create register-friendly variable
      char *src_ptr = source_.data();

      do {
        int  ch;
        auto eqc = ch_to_eqc[(ch = *src_ptr++)];

        state_ = next_state(state_, eqc);
        tok.len += lex_inside[utype(state_)];

        switch (state_) {
        case LexicalState::eNextChar: {
          ++tok.col_num;
          break;
        }
        case LexicalState::eNewLine: {
          ++tok.line_num;
          tok.col_num = 1;
          break;
        }
        case LexicalState::eIdentifier: {
          tok.kind = TokenKind::eIdentifier;
          break;
        }
        case LexicalState::eOperatorMultiEnd:
        case LexicalState::eIdentifierEnd: {
          tok.update_repr(src_ptr, 1, 0);
          tok.to_hyponym();

          // End of token requires reevaluation
          src_ptr += ch_reeval[ch];
          // TODO(rihtwis-weard): update col pos according to token length if
          //                    reevaluation occurred

          tok.reset();
          tokens_.push_back(tok);
          break;
        }
        case LexicalState::eNumeral: {
          tok.kind = TokenKind::eNumeral;
          break;
        }
        case LexicalState::eNumeralEnd: {
          tok.update_repr(src_ptr, 1, 0);
          src_ptr += ch_reeval[ch];
          tok.reset();
          tokens_.push_back(tok);
          break;
        }
        case LexicalState::eSeparator: {
          tok.kind = TokenKind::eSeparator;
          tok.update_repr(src_ptr, 0, 0);

          tok.reset();
          tokens_.push_back(tok);
          break;
        }
        case LexicalState::eOperatorSingle: {
          tok.kind = TokenKind::eOperator;
          tok.update_repr(src_ptr, 0, 0);

          tok.reset();
          tokens_.push_back(tok);
          break;
        }
        default:
          break;
        }

      } while (not should_exit(state_));

      return tokens_;
    }
  };

} // namespace claire
