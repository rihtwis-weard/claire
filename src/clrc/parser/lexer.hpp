#pragma once

#include <vector>

#include "../exception.hpp"
#include "state_machine.h"
#include "token.h"

namespace claire::parser {

  class Lexer {
    LexicalState       state_;
    std::string        source_;
    std::vector<Token> tokens_;

  public:
    explicit Lexer(std::string const &fpath);

    std::vector<Token> lex();
  };

} // namespace claire::parser
