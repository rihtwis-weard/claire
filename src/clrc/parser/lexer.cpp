#include "lexer.hpp"

namespace claire::parser {

  Lexer::Lexer(std::string const &fpath)
    : state_{LexicalState::eNextChar}
    , source_{loads(fpath)} {
  }

  std::vector<Token> Lexer::lex() {
    Token tok{};

    // Alias pointer to buffer to create register-friendly variable
    char *src_ptr = source_.data();

    do {
      unsigned char ch;
      auto          eqc = ch_to_eqc[(ch = *src_ptr++)];

      state_ = next_state(state_, eqc);
      tok.len += lex_inside[utype(state_)];

      switch (state_) {
      case LexicalState::eNextChar: {
        ++tok.col_num;
        break;
      }
      case LexicalState::eNewLine: {
        ++tok.line_num;
        tok.col_num = 1;
        break;
      }
      case LexicalState::eIdentifier: {
        tok.kind = TokenKind::eIdentifier;
        break;
      }
      case LexicalState::eOperatorMultiEnd:
      case LexicalState::eIdentifierEnd: {
        tok.update_repr(src_ptr, 1, 0);
        tok.to_hyponym();

        // End of token requires reevaluation
        src_ptr += ch_reeval[ch];
        // TODO(rihtwis-weard): update col pos according to token length if
        //                    reevaluation occurred

        tok.reset();
        tokens_.push_back(tok);
        break;
      }
      case LexicalState::eString: {
        tok.kind = TokenKind::eStringLiteral;
        break;
      }
      case LexicalState::eStringEnd: {
        tok.update_repr(src_ptr, 1, 1);
        src_ptr += ch_reeval[ch];
        tok.reset();
        tokens_.push_back(tok);
        break;
      }
      case LexicalState::eNumeral: {
        tok.kind = TokenKind::eNumeral;
        break;
      }
      case LexicalState::eNumeralEnd: {
        tok.update_repr(src_ptr, 1, 0);
        src_ptr += ch_reeval[ch];
        tok.reset();
        tokens_.push_back(tok);
        break;
      }
      case LexicalState::eSeparator: {
        tok.kind = TokenKind::eSeparator;
        tok.update_repr(src_ptr, 0, 0);

        tok.reset();
        tokens_.push_back(tok);
        break;
      }
      case LexicalState::eOperatorSingle: {
        tok.update_repr(src_ptr, 0, 0);
        tok.to_hyponym(TokenKind::eOperator);

        tok.reset();
        tokens_.push_back(tok);
        break;
      }
      case LexicalState::eEOF: {
        throw unexpected_eof{};
      }
      case LexicalState::eError: {
        throw invalid_lexeme{};
      }
      default:
        break;
      }

    } while (not should_exit(state_));

    return tokens_;
  }

} // namespace claire::parser
