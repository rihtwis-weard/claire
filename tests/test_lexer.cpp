#define APPROVALS_UT
#include "ApprovalTests.hpp"

#include "lexer.hpp"

int main() {
  using namespace boost::ut;
  using namespace ApprovalTests;

  auto directory = Approvals::useApprovalsSubdirectory("golden_files");

  "hello_world"_test = []() {
    auto lexemes = claire::Lexer{"../../examples/hello_world.clr"}.lex();
    Approvals::verifyAll("hello_world.clr", lexemes);
  };

  "fib"_test = []() {
    auto lexemes = claire::Lexer{"../../examples/fib.clr"}.lex();
    Approvals::verifyAll("fib.clr", lexemes);
  };

  "exception"_test = []() {
    std::string sources[]{"unsupported_multi_operator.clr"};

    for (auto const &src : sources) {
      auto lexer = claire::Lexer{"../../tests/data/" + src};
      expect(throws([&]() { lexer.lex(); })) << "throws any exception";
    }
  };
}
