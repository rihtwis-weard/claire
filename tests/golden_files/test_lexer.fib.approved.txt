fib.clr


[0] =     Reserved | let |
[1] =   Identifier | fib |
[2] =    Separator | \ |
[3] =   Identifier | x |
[4] =    Separator | : |
[5] =   Identifier | Integer |
[6] =     Operator | = |
[7] =     Reserved | if |
[8] =   Identifier | x |
[9] =     Operator | < |
[10] =      Numeral | 3 |
[11] =      Numeral | 1 |
[12] =     Reserved | else |
[13] =   Identifier | fib |
[14] =     Operator | ( |
[15] =   Identifier | x |
[16] =     Operator | - |
[17] =      Numeral | 1 |
[18] =     Operator | ) |
[19] =     Operator | + |
[20] =   Identifier | fib |
[21] =     Operator | ( |
[22] =   Identifier | x |
[23] =     Operator | - |
[24] =      Numeral | 2 |
[25] =     Operator | ) |
[26] =      Numeral | 40 |
[27] =     Operator | |> |
[28] =   Identifier | fib |
[29] =     Operator | ( |
[30] =     Operator | ) |
[31] =     Operator | |> |
[32] =   Identifier | IO |
[33] =     Operator | . |
[34] =   Identifier | puts |
[35] =     Operator | ( |
[36] =     Operator | ) |

